[WIP] CLARIAH Requirements for infrastructure and Software/Services (#5)
CLARIAH/clariah-plus: docs/requirements/software-requirements.md

# CLARIAH Technical Requirements for Software and Services

## Introduction

This document intends to specify minimal technical requirements for CLARIAH
software and software services. It is aimed at developers and external
evaluators. It provides a minimal list of specific and directly actionable criteria to hold the software
against. The aim is to ensure a certain level of software quality and
sustainability, and to lay out basic requiments needed for interoperability
within the larger CLARIAH infrastructure.

> 

Reviewer: Software quality and sustainability, important as they are, are no goals in themselves. Perhaps we can relate these goals to the _use_ of software, just as we do for the infrastructure requirements:

```suggestion
Only software that is built and maintained well will get used and thus provide value.

```

The document is loosely derived from [earlier work to establish Software Quality

> Guidelines](https://github.com/CLARIAH/software-quality-guidelines/) within CLARIAH-CORE, but attempts to condense this

Reviewer: Use fewer words:

```suggestion
Guidelines](https://github.com/CLARIAH/software-quality-guidelines/) within CLARIAH-CORE, but condenses that
```

elaborate work to the most relevant and directly actionable points.

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”,
“SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be
interpreted as described in [RFC
2119](https://datatracker.ietf.org/doc/html/rfc2119).

## Basic Software Requirements

### 1. The software's source code *MUST* be stored in a public **version control system** (VCS).

> 1. The use of a publicly accessible VCS platforms such as Github, Gitlab, Bitbucket or SourceHut is *REQUIRED*. This may be self-hosted by an institution rather than outsourcing it to a third party.

Reviewer: Just examples, so mention the two most common ones only:

```suggestion
1. The use of a publicly accessible VCS platforms such as GitHub or GitLab is *REQUIRED*. This may be self-hosted by an institution rather than outsourcing it to a third party.
```

Author: You're probably right.. it's just that I like SourceHut a lot, despite it being smaller :) It also shows that we don't strictly need to adhere to the exact same paradigms GitHub/gitlab/bitbucket use (SourceHut is more e-mail based). So I'll keep it as-is for now.

2. CLARIAH is registered as an [organization](https://github.com/CLARIAH) on github, any CLARIAH software project *MAY* be hosted there.
3. Git is *RECOMMENDED*
4. All content, including documentation, *SHOULD* be stored alongside the code in the VCS.

> 5. The source code repository *SHOULD* be publically accessible from day 1 of development; it is *MUST NOT* be reduced to a release/file-transfer mechanism for completed products. Exceptions are when things are under embargo or privacy-sensitive.

Reviewer: Give reasons:

```suggestion
5. The source code repository *SHOULD* be publicly accessible from day 1 of development so interested third parties can follow development and provide feedback. The software code *MUST NOT* be reduced to a release/file-transfer mechanism for completed products. Exceptions are when code is under embargo or privacy-sensitive.
```

### 2. A README file *MUST* be provided in the root directory of the VCS

> 1. This should be a readable plain-text files. Markup formats such as markdown (``README.md``, or alternatively ReStructuredText, ``README.rst``) are *recommended*. Other formats such as LaTeX, HTML, Word, etc *MUST NOT* be used for the README.

Reviewer: ```suggestion
1. This should be a readable plain-text files. Markup formats such as Markdown (``README.md``, or alternatively ReStructuredText, ``README.rst``) are *RECOMMENDED*. Other formats such as LaTeX, HTML, Word, etc *MUST NOT* be used for the README.
```

2. The README *MUST* have a clear description of what the software does (what problems it solves) and whom it is intended for.

> 3. The README *MUST* make the current status of the software clear, is it ready for production, experimental, or a proof-of-concept? It *MUST* also make clear whether the software is actively maintained or not. The use of the [repostatus](https://www.repostatus.org/) vocabulary is *RECOMENDED* to this end. A simple repostatus badge suffices.

Reviewer: ```suggestion
3. The README *MUST* make the current status of the software clear: is it ready for production, experimental, or a proof-of-concept? It *MUST* also make clear whether the software is actively maintained or not. The use of the [repostatus](https://www.repostatus.org/) vocabulary is *RECOMENDED* to this end. A simple repostatus badge suffices.
```

> 4. The README *MUST* make clear who a) wrote the software and have a contact link, b) acknowledge the funders
> https://github.com/IdentityPython/SATOSA5. The README *MUST* provide (or link to) installation instructions.

Reviewer: Probably a (paste) mistake?

```suggestion
4. The README *MUST* make clear wrote the software, including a contact link, and acknowledge the funders. 
5. The README *MUST* provide (or link to) installation instructions.
```

6. The README *MUST* provide (or link to) usage instructions for a quick start, explaining how the first task can be performed with the system.

> 7. The README *should* make explicit any software and minimal hardware requirements the software relies on. This includes making explicit the software's main dependencies, target OS, and minimally required resources.

Reviewer: ```suggestion
7. The README *SHOULD* make explicit any software and minimal hardware requirements the software relies on. This includes making explicit the software's main dependencies, target OS, and minimally required resources.
```

8. The README *MUST* make clear (link to) a public support channel (e.g. an issue tracker, mailing list) where people can submit bug reports and feature requests.

### 3. The software *MUST* be distributed as open source under an OSI approved license

The full license *MUST* be stored in a ``LICENSE`` (``LICENSE.md``) file in the root of the VCS, the same format
recommendations and restrictions apply as to the ``README`` file.

Developers *MUST* ensure the license is not in conflict with the licenses of the dependencies they use.

* See the list of [OSI-approved licenses](https://opensource.org/licenses)

### 4. The software *MUST* be released periodically with clear version numbers

> Periodic releases of the software should be released with a clear version number.

Reviewer: ```suggestion
Periodic releases of the software *SHOULD* be released with a clear version number.
```

> 1. The use of [semantic versioning](https://semver.org/) is *strongly recommended*.

Reviewer: ‘Strongly recommended’ is not a RFC 2119 keyword.

Author: Right, I'll change it to "RECOMMENDED"

   split into reusable software libraries/tools rather than be part of an
   indivisible monolithical whole. This ensures the work can be reused where
   appropriate.
2. Software *MUST* separate code from configuration, they *MUST NOT* be hard-coded to only work on a single system or
   for a specific user; hard-coded user/installation-specific paths and sensitive credentials *MUST NOT* be part of the
   software source code. Software *MUST* be made configurable (through configuration files, command line parameters,
   environment variables) to be deployable in multiple environments.

### 8. Software *SHOULD* come with automated tests

1. To ensure software quality, software *SHOULD* have a test set consisting of
unit and/or integration tests with a fair degree of coverage.
2. Automatic *Continuous Integration* infrastructure such as GitHub Actions, GitLab CI/CD,
Jenkins or others *SHOULD* be used to automatically test the software upon any
code change.
3. Additionaly, code coverage tests and dependency version checks
*MAY* be used.

Deviations from this are only acceptable for initial proof-of-concept or highly
experimental software (see 1.3).

> ### 9. Software **MUST** define codemeta software metadata along with the source code

Reviewer: ```suggestion
### 9. Software **MUST** define CodeMeta software metadata along with the source code
```

   mounted volume) (See point 3 of the [Infrastructure Requirements (IR)](infrastructure-requirements.md))

### 13. Service developers *SHOULD* provide an initial template when of multi-container orchestration is needed

If a complex service consists of multiple interacting containers, the developers *SHOULD* provide an initial template in
the form of a Docker Compose configuration or a Kubernetes deployment configuration that illustrates how the containers
are orchestrated to form the application. This *MUST* be maintained in a VCS repository (infrastructure as code
principle). Infrastructure operators can build on this example to deploy the application.

### 14. Services *MUST* be compatible with CLARIAH's authentication and authorization infrastructure

All services open to end-users and which require some form of user authentication should be compatible with
CLARIAH's authentication and authorization infrastructure. That is, they should be able to communicate with CLARIAH's
[SATOSA](https://github.com/IdentityPython/SATOSA) Authentication Provider. It is *RECOMMENDED* to use OpenID Connect

> for this communication. Instruction can be found [(TODO add link!)]().

Reviewer: ```suggestion
for this communication. Instruction can be found [here](https://github.com/CLARIAH/IG-DevOps/tree/main/docs/authentication).
```

### 15. Services *MUST* expose a public endpoint providing their specification

1. Swagger, WADL, and CLAM are *SUGGESTED* as possible interface description languages.
2. The endpoint *SHOULD NOT* be hindered by any authentication barriers.

> ### 16. Services *SHOULD* expose a public endpoint providing high-level codemeta metadata

Reviewer: ```suggestion
### 16. Services *SHOULD* expose a public endpoint providing high-level CodeMeta metadata
```

Is any standard URL for this endpoint described by CodeMeta (cf. the `/metrics` endpoint for Prometheus/OpenTelemetry)?

Author: No, CodeMeta doesn't define any of that, we'll have to make a choice ourselves.

This is the corollary of point 9 for software. The codemeta metadata for the service as a whole must be available at a

> public endpoint returning a codemeta JSON-LD file. Note that this is not does describe the API (unlike point 14). This

Reviewer: ```suggestion
public endpoint returning a codemeta JSON-LD file. Note that this does describe the API (unlike point 14). This
```

Author: codemeta does NOT describe the API, I fixed the typo accordingly now.

   mounted volume) (See point 3 of the [Infrastructure Requirements (IR)](infrastructure-requirements.md))

### 15. Service developers *SHOULD* provide an initial template when multi-container orchestration is needed

If a complex service consists of multiple interacting containers, the developers *SHOULD* provide an initial template in
the form of a Docker Compose configuration or a Kubernetes deployment configuration that illustrates how the containers
are orchestrated to form the application. This *MUST* be maintained in a VCS repository (infrastructure as code
principle). Infrastructure operators can build on this example to deploy the application.

### 17. Services *MUST* be compatible with CLARIAH's authentication and authorization infrastructure

All services open to end-users and which require some form of user authentication *MUST* be compatible with
CLARIAH's authentication and authorization infrastructure. That is, they should be able to communicate with CLARIAH's
[SATOSA](https://github.com/IdentityPython/SATOSA) Authentication Provider. It is *RECOMMENDED* to use OpenID Connect

> for this communication. Instruction can be found [here](https://github.com/CLARIAH/IG-DevOps/tree/main/docs/authentication).

Reviewer: From https://github.com/CLARIAH/IG-DevOps/blob/main/docs/authentication/authentication_clariah_nl.md:

> all of the identity providers registered at CLARIN (see "Participating Identity Federations" at CLARINs Service Provider Federation). This basically means: all European academic institutions. In addition to this CLARIN-provided collection of identity providers, Beeld en Geluid is also available as an identity provider in discovery.clariah.nl.

So this requirement makes a lot of sense for applications geared towards academic users. But what about applications with other types of users? Case in point: NDE’s [Solid Collection Registration System](https://github.com/netwerk-digitaal-erfgoed/solid-crs). What would be the added value for this type of applications in using SATOSA?

So should we:

- generalise this requirement to be about OIDC-compatibility on the part of the application, so it can be connected to _any_ OIDC IdP?
- or document more clearly how to use CLARIAH auth with non-academic users (I guess BenG is an example of this)?
